lib/Parse/ParseDecl.cpp

```diff
@@ -5767,27 +5767,32 @@ Parser::parseDeclEnumCase(ParseDeclOptions Flags,
       SourceLoc TokLoc = Tok.getLoc();
       StringRef TokText = Tok.getText();
 
-      // For recovery, see if the user typed something resembling a switch
-      // "case" label.
-      llvm::SaveAndRestore<decltype(InVarOrLetPattern)>
-      T(InVarOrLetPattern, Parser::IVOLP_InMatchingPattern);
-      parseMatchingPattern(/*isExprBasic*/false);
+      if (!NameIsKeyword || Tok.isIdentifierOrUnderscore()) {
+        // For recovery, see if the user typed something resembling a switch
+        // "case" label.
+        llvm::SaveAndRestore<decltype(InVarOrLetPattern)>
+        T(InVarOrLetPattern, Parser::IVOLP_InMatchingPattern);
+        parseMatchingPattern(/*isExprBasic*/false);
+      }
```

Comment:
I think this was trying to diagnose something like:
```
enum E {
  case let .foo(x, y):
}
```
In this case case, `NameIsKeyword` is true, but we want to diagnose this as just `error: 'case' label can only appear inside a 'switch' statement`.

Reply:
Humm.. that makes sense, seems like the pattern matching only needs to be done for certain keywords like var, let and is. So make sense adds that there too? e.g. 
```swift
if (!NameIsKeyword
          || Tok.isIdentifierOrUnderscore()
          || Tok.isAny(tok::kw_var, tok::kw_let, tok::kw_is)) {
// pattern matching
}
```

Reply:
But acctually adding that will cause an `expected pattern` error when 
```swift
enum E {
  case identifier
  case var 
  case identifier2
}
```
So, I'm not sure what would be a condition to try this matching.

Reply:
How about using `BacktrackingScope`? Like:
```cpp
if (Tok.is(tok::identifier)) {
    // ...
} else { // line 5764.
  {
    // Check if it's 'case' statement outside of 'switch'.
    BacktrackingScope backtrack(*this);
    llvm::SaveAndRestore<decltype(InVarOrLetPattern)> T(
        InVarOrLetPattern, Parser::IVOLP_InMatchingPattern);
    parseMatchingPattern(/*isExprBasic*/false);
    if (confumeIf(tok::colon)) {
      // If we see ':' after pattern, it's probably a 'case' stmt outside 'switch'.
      backtrack.cancel()
      diagnose(CaseLoc, diag::case_outside_of_switch, "case");
      Status.setIsParseError();
      return Status;
    }
    // Otherwise, backtrack to the position after 'case'.
  }

  if (Tok.is(tok::keyword) && !Tok.isAtStartOfLine()) {
    // If it's a keyword on the same line, diagnose and consume it as if it's escaped.
    auto TokText = Tok.getText();
    NameLoc = consumeToken();
    diagnose(NameLoc, diag::keyword_cant_be_identifier, TokText);
    diagnose(NameLoc, diag::backticks_to_escape)
        .fixItReplace(NameLoc, "`" + TokText + "`");
    Name = Context.getIdentifier(TokText);
  } else if (CommaLoc.isValid()) {
    // If it's a random token after comma, bailout.
    diagnose(Tok, diag::expected_identifier_after_case_comma);
    Status.setIsParseError();
    return Status;
  } else {
    diagnose(CaseLoc, diag::expected_identifier_in_decl, "enum 'case'");
    NameLoc = CaseLoc;
  }
}
```

Reply:
@rintaro thank's for the suggestion, I didn't know `BacktrackingScope` was a thing. 
Just updated the code 👍